packages/tooling/fast-tooling-react/src/navigation/navigation-tree-item.tsx

```diff
@@ -13,113 +13,165 @@ import {
+            value={value}
+            onChange={handleChange}
+            onBlur={handleBlur}
+        />
+    );
+}
+
+function treeItem(
     handleClick: React.MouseEventHandler<HTMLElement>,
+    handleChange: React.ChangeEventHandler<HTMLInputElement>,
+    handleBlur: React.FocusEventHandler<HTMLInputElement>,
     handleExpandClick: React.MouseEventHandler<HTMLElement>,
     handleKeyDown: React.KeyboardEventHandler<HTMLElement>,
+    isEditable: boolean,
+    isCollapsible: boolean,
     className: string,
     expandTriggerClassName: string,
     contentClassName: string,
-    children: React.ReactNode,
+    displayTextInputClassName: string,
+    text: string,
     dictionaryId: string,
     navigationConfigId: string,
     ref?: (node: HTMLAnchorElement) => React.ReactElement<any>
 ): React.ReactElement {
+    const displayText: React.ReactNode = isEditable
+        ? editableOverlay(displayTextInputClassName, text, handleChange, handleBlur)
+        : isCollapsible
+        ? treeItemCollapsible(
+              handleClick,
+              handleKeyDown,
+              contentClassName,
+              text,
+              dictionaryId,
+              navigationConfigId,
+              ref
+          )
+        : treeItemEndLeaf(
+              handleClick,
+              handleKeyDown,
+              contentClassName,
+              text,
+              dictionaryId,
+              navigationConfigId,
+              ref
+          );
+
     return (
         <span className={className}>
             <span className={expandTriggerClassName} onClick={handleExpandClick} />
-            <a
-                className={contentClassName}
-                onClick={handleClick}
-                ref={ref}
-                tabIndex={0}
-                onKeyDown={handleKeyDown}
-                data-dictionaryid={dictionaryId}
-                data-navigationconfigid={navigationConfigId}
-            >
-                {children}
-            </a>
+            {displayText}
         </span>
     );
 }
 
+function treeItemEndLeaf(
+    handleClick: React.MouseEventHandler<HTMLElement>,
+    handleKeyDown: React.KeyboardEventHandler<HTMLElement>,
+    contentClassName: string,
+    text: string,
+    dictionaryId: string,
+    navigationConfigId: string,
+    ref?: (node: HTMLAnchorElement) => React.ReactElement<any>
+): React.ReactElement {
+    return (
+        <a
+            className={contentClassName}
+            onClick={handleClick}
+            ref={ref}
+            tabIndex={0}
+            onKeyDown={handleKeyDown}
+            data-dictionaryid={dictionaryId}
+            data-navigationconfigid={navigationConfigId}
+        >
+            {text}
+        </a>
+    );
+}
+
 function treeItemCollapsible(
     handleClick: React.MouseEventHandler<HTMLElement>,
-    handleExpandClick: React.MouseEventHandler<HTMLElement>,
     handleKeyDown: React.KeyboardEventHandler<HTMLElement>,
-    className: string,
```

Comment:
You might consider just making these functional components; You'll get some of these props for free w/o enumeration and prop order won't matter.

Reply:
I think that's out of scope for this PR, but I've created an [issue](https://github.com/microsoft/fast/issues/4386) for it.